Firmware/PlatformIO/main_mcu/src/bms.cpp

```diff
@@ -19,6 +19,39 @@ void BMS::Tick(std::chrono::milliseconds elapsed_time)
     // todo
 }
 
+// Find maximum discharge and regen current
+void BMS::CalculateSOE()
+{
+    float currentPerCell = current[0] / kNumCellsParallel;
+    float internalResistancePerSeriesElement = kInternalResistance / kNumCellsParallel;
+
+    // Find highest and lowest open circuit voltage
+    float minOpenCircuitVoltage = minVoltage + (currentPerCell * kInternalResistance);
+    float maxOpenCircuitVoltage = maxVoltage + (currentPerCell * kInternalResistance);
+
+    // Limit at V_open + I * (R_internal / numCellsParallel) = V_boundary
+    //  => I = (numCellsParallel / R_internal) * (V_boundary - V_open)
+    float maxDischargeVoltage = minOpenCircuitVoltage - kCellUndervoltage;
+    float maxRegenVoltage = kCellOvervoltage - maxOpenCircuitVoltage;
+    float uncappedDischargeCurrent = maxDischargeVoltage / internalResistancePerSeriesElement;
+    float uncappedRegenCurrent = maxRegenVoltage / internalResistancePerSeriesElement;
+
+    // I = P / V
+    float powerCappedDischargeCurrent = kMaxPowerOutput / maxDischargeVoltage;
+    float powerCappedRegenCurrent = kMaxPowerOutput / maxRegenVoltage;
```

Comment:
These should use the full voltage of the pack/the sum of all of the cell voltages (which can be calculated using std::accumulate on the array of cell voltages)